### PR TITLE
Macros for generating cmd*.rb , conf/*.yaml files 

### LINKED ISSUE #72: Add cmake helper functions for configuring / generating cmd / conf files
## Desired behavior

We currently have a lot of repeated boilerplate code in each library that generates the `conf/*.yaml` and `cmd*.rb` files. It would be nice to provide some cmake helper functions. We could install that with `ign-tools` since it is related to this package.

For example in ign-transport:

* https://github.com/ignitionrobotics/ign-transport/blob/ignition-transport11_11.0.0/conf/CMakeLists.txt
* https://github.com/ignitionrobotics/ign-transport/blob/ignition-transport11_11.0.0/src/cmd/CMakeLists.txt#L63-L108

and also in ign-plugin

* https://github.com/ignitionrobotics/ign-plugin/blob/0da6f078a3b98c3ac962378272dd29c80b70b119/loader/conf/CMakeLists.txt
* https://github.com/ignitionrobotics/ign-plugin/blob/0da6f078a3b98c3ac962378272dd29c80b70b119/loader/src/cmd/CMakeLists.txt

The helper functions should be able to generate separate files for installation and testing.

## Alternatives considered

* We could add cmake helper functions to ign-cmake.
* We could continue with the current approach that involves lots of duplicate code.

### PR DESCRIPTION
**Pull request by harshmahesheka**
# 🎉 New feature

Closes #72 

## Summary
Added Macros for generating cmd*.rb and conf/*.yaml files which decreases boilerplate code from repositories.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers
## Test It
I have replaced the boilerplate code with macros in my fork [here](https://github.com/harshmahesheka/gz-transport/tree/test_pr). You can build this and confirm that the correct cmd and conf are being generated and installed.  

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: cmake/helper_functions.cmake [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      "${cmd_script_configured}"
+      @ONLY)
+
+    file(GENERATE
+      OUTPUT "${cmd_script_generated}"
+      INPUT  "${cmd_script_configured}")
+
+    # Install the ruby command line library in an unversioned location.
+    install(FILES ${cmd_script_generated} DESTINATION lib/ruby/ignition)
+endmacro()
```
Comment (arjo129) [inline cmake/helper_functions.cmake:70]: Knit: Add newline at end of file.

File: cmake/helper_functions.cmake [hunk 2]
```diff
@@ -0,0 +1,70 @@
+# ign_generate_conf()
+# Generate conf/*.yaml files for projects
```
Comment (arjo129) [inline cmake/helper_functions.cmake:2]: Knit: can we name this file something different. "helper_functions is not very descriptive.

File: cmake/helper_functions.cmake [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+macro(ign_generate_conf)
```
Comment (arjo129) [inline cmake/helper_functions.cmake:3]: Meybe change the name of this macro?

File: cmake/helper_functions.cmake [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    configure_file(
+      "cmd${IGN_DESIGNATION}.rb.in"
+      "${cmd_script_configured_test}"
+      @ONLY)
+
+    file(GENERATE
+      OUTPUT "${cmd_script_generated_test}"
+      INPUT  "${cmd_script_configured_test}")
+
```
Comment (arjo129) [inline cmake/helper_functions.cmake:49]: Knit: remove excess spaces.

**Review by arjo129 — CHANGES_REQUESTED** [review]
So I've done a cursory pass of the macros and there are some minor knits I've pointed out. The bigger question is have you used these macros anywhere? How can I test that they run correctly? If you'd like you can create forks of the libraries and point me to branches where you are using them.

As a good rule of thumb, always include instructions on how to test your code in a separate PR.

File: cmake/helper_functions.cmake (hunk 2 above)
Comment (harshmahesheka) [reply cmake/helper_functions.cmake]: How about ign_generate_files, we can keep other macros that generate boilerplate under this later.

File: cmake/helper_functions.cmake (hunk 3 above)
Comment (harshmahesheka) [reply cmake/helper_functions.cmake:3]: How about ign_generate_conf_files, or could you suggest a better name?

**Comment by harshmahesheka** [comment]
> So I've done a cursory pass of the macros and there are some minor knits I've pointed out. The bigger question is have you used these macros anywhere? How can I test that they run correctly? If you'd like you can create forks of the libraries and point me to branches where you are using them.
> 
> As a good rule of thumb, always include instructions on how to test your code in the PR description.

I have resolved all the knits and added instructions for testing it with ign-transport.I will try to think of some better names

File: cmake/helper_functions.cmake [hunk 5]
```diff
@@ -0,0 +1,68 @@
+#ign_generate_conf()
+#Generate conf/*.yaml files for projects
+macro(ign_generate_conf)
+    #Used only for internal testing.
+    set(ign_library_path "${CMAKE_BINARY_DIR}/test/lib/$<CONFIG>/ruby/ignition/cmd${IGN_DESIGNATION}${PROJECT_VERSION_MAJOR}")
+
+    #Generate a configuration file for internal testing.
```
Comment (arjo129) [inline cmake/helper_functions.cmake:7]: Eh, comments should have a space between the `#` and the body. By remove space i meant the extra lines.

File: cmake/helper_functions.cmake (hunk 3 above)
Comment (harshmahesheka) [reply cmake/helper_functions.cmake]: I have changed the name to configuration_functions as discussed.

File: cmake/helper_functions.cmake (hunk 2 above)
Comment (harshmahesheka) [reply cmake/helper_functions.cmake]: I have changed the name to configuration_functions as discussed.

File: cmake/helper_functions.cmake (hunk 3 above)
Comment (harshmahesheka) [reply cmake/helper_functions.cmake]: Let's keeping the same name as we couldn't come up with a better name

**Comment by harshmahesheka** [comment]
As discussed I am closing this pr and opening another one here https://github.com/gazebosim/gz-cmake/pull/270

**Pull request closed without merging**